enhancements/insights/conditional-data-gathering.md

```diff
@@ -0,0 +1,199 @@
+and image streams of the same namespace when alert `SamplesImagestreamImportFailing` is firing
+and add it to the next archive.
+
+```json
+[{
+  "conditions": [{
+    "type": "alert_is_firing",
+    "params": {
+      "name": "SamplesImagestreamImportFailing"
+    } 
+  }],
+  "gathering_functions": {
+    "logs_of_namespace": {
+      "namespace": "openshift-cluster-samples-operator",
+      "tail_lines": 100
+    },
+    "image_streams_of_namespace": {
+      "namespace": "openshift-cluster-samples-operator"
+    }
+  }
+}]
+```
+
+Conditions can have type and parameters, implemented types are:
+
+- `alert_is_firing` which is met when the alert with the name from parameter `name` is firing
+
+A gathering function consists of its name and parameters, implemented functions are:
+
+- `logs_of_namespace` which collects logs from each container in the namespace from parameter `namespace`
+limiting it to only last N lines from parameter `tail_lines`
+- `image_streams_of_namespace` which collects image streams of the namespace from parameter `namespace`
+
+The proposal is to implement the next process of adding new rules:
+
+1. We have a repo with json configs defining the rules. The repo will have some simple CI with validation
+against JSON schema and possibly some review process. 
+We have created a PoC for the repo: 
+
+https://github.com/tremes/insights-operator-gathering-rules
+
+or the version with an example of JSON schema validation:
+
+https://github.com/tremes/insights-operator-gathering-rules/tree/schema 
+
+2. There's a service living in cloud.redhat.com fetching the rules from the repo and providing them through its API.
+The very first version is going to provide just all the rules, but later we may consider splitting them by 
+cluster version and introducing some more complicated logic around fetching the rules
+3. Insights Operator fetches a config with the rules from the service and unmarshalls JSON to Go structs
+4. Insights Operator makes the most important validation which checks the next things:
+
+- The JSON version of the config matches the structs defined in the code 
+- For each rule, there's at least one gathering function
+- Only implemented conditions can be used
+- Alert name from `alert_is_firing` condition should be a string of length between 1 and 128 
+and consist of only alphanumeric characters
+- Only implemented gathering functions can be used
+- Namespace from `gather_logs_of_namespace` function should be a string of length between 1 and 128, 
+match the next regular expression `^[a-zA-Z]([a-zA-Z0-9\-]+[\.]?)*[a-zA-Z0-9]$` and start with `openshift-`
+- Tail lines from `gather_logs_of_namespace` function should be an integer between 1 and 4096
+- Namespace from `gather_image_streams_of_namespace` function should be a string of length between 1 and 128, 
+match the next regular expression `^[a-zA-Z]([a-zA-Z0-9\-]+[\.]?)*[a-zA-Z0-9]$` and start with `openshift-`
```

Comment:
Not sure but these are maybe impl details from the perspective of this document.

Reply:
I think it's better to keep it, maybe move to another sections, but it should be there cuz it's important to understand how safe the solution is

Reply:
are we ok to resolve this - i.e. keep it in the doc ... @tremes ?

Reply:
I am fine with resolving this conversation.